File: sdk/ruby/test/wire/context_sets_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class ContextSetsWireTest < WireMockTestCase
  def setup
    super

    @client = AgentTrunk::Client.new(
      access_token: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_context_sets_sources_with_wiremock
    test_id = "context_sets.sources.0"

    @client.context_sets.sources(request_options: {
      additional_headers: {
        "X-Test-Id" => "context_sets.sources.0"
      }
    })

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/context-sources",
      query_params: nil,
      expected: 1
    )
  end

  def test_context_sets_list_with_wiremock
    test_id = "context_sets.list.0"

    @client.context_sets.list(request_options: {
      additional_headers: {
        "X-Test-Id" => "context_sets.list.0"
      }
    })

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/context-sets",
      query_params: nil,
      expected: 1
    )
  end

  def test_context_sets_create_with_wiremock
    test_id = "context_sets.create.0"

    @client.context_sets.create(
      name: "name",
      sources: [{
        source_trunk_id: "sourceTrunkId",
        source_scope_id: "sourceScopeId",
        environment_id: "environmentId",
        context_key: "contextKey",
        revision_id: "revisionId",
        package_digest: "packageDigest",
        mount_path: "mountPath",
        required: true
      }],
      request_options: {
        additional_headers: {
          "X-Test-Id" => "context_sets.create.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/context-sets",
      query_params: nil,
      expected: 1
    )
  end

  def test_context_sets_resolve_with_wiremock
    test_id = "context_sets.resolve.0"

    @client.context_sets.resolve(
      context_set_id: "contextSetId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "context_sets.resolve.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/context-sets/contextSetId/resolve",
      query_params: nil,
      expected: 1
    )
  end
end

File: sdk/rust/tests/context_sets_test.rs
use agenttrunk::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_context_sets_sources_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentTrunk::new(config).expect("Failed to build client");

    let result = client
        .context_sets
        .sources(
            &SourcesQueryRequest {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("GET", "/v1/context-sources", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_context_sets_list_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentTrunk::new(config).expect("Failed to build client");

    let result = client
        .context_sets
        .list(
            &ContextSetsListQueryRequest {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("GET", "/v1/context-sets", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_context_sets_create_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentTrunk::new(config).expect("Failed to build client");

    let result = client
        .context_sets
        .create(
            &CreateContextSetsRequest {
                name: "name".to_string(),
                sources: vec![ContextSetSource {
                    source_trunk_id: "sourceTrunkId".to_string(),
                    source_scope_id: "sourceScopeId".to_string(),
                    environment_id: "environmentId".to_string(),
                    context_key: "contextKey".to_string(),
                    revision_id: "revisionId".to_string(),
                    package_digest: "packageDigest".to_string(),
                    mount_path: "mountPath".to_string(),
                    required: true,
                    ..Default::default()
                }],
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/context-sets", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_context_sets_resolve_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentTrunk::new(config).expect("Failed to build client");

    let result = client
        .context_sets
        .resolve(&"contextSetId".to_string(), None)
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/context-sets/contextSetId/resolve", None, 1)
        .await
        .unwrap();
}
